BUG: KeyError: 'GPU_BACKEND

With a CUDA install of `pykokkos-base` and running `python runtests.py -t tests/test_views.py ` I see failures like this:

```
_________________________________________________________________________________________________________________________________ TestViews.test_v3d _________________________________________________________________________________________________________________________________

self = <tests.test_views.TestViews testMethod=test_v3d>

    def setUp(self):
        self.threads: int = 10
        self.i_1: int = 10
        self.i_2: int = 15
        self.i_3: int = 20
        self.i_4: int = 10
    
>       self.functor = ViewsTestFunctor(self.threads, self.i_1, self.i_2, self.i_3, self.i_4)

tests/test_views.py:192: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/test_views.py:96: in __init__
    self.cp_view: pk.View2D[int] = pk.from_cupy(cp_arr)
pykokkos/interface/views.py:614: in from_cupy
    if km.get_gpu_framework() == "Cuda":
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

    def get_gpu_framework() -> str:
        """
        Get the framework used by the GPU
    
        :returns: the framework as a string
        """
    
>       return CONSTANTS["GPU_BACKEND"]
E       KeyError: 'GPU_BACKEND'

pykokkos/kokkos_manager/__init__.py:198: KeyError

```

Maybe pykokkos-base/Kokkos aren't setup to deal with the consumer card and driver scenario I have on this desktop machine at work when calling `get_device_available` under the hood?. For example, I seem to be able to circumvent the issue with this diff and still run on the GPU:

```diff
diff --git a/pykokkos/interface/views.py b/pykokkos/interface/views.py
index ba3f3ba..3066959 100644
--- a/pykokkos/interface/views.py
+++ b/pykokkos/interface/views.py
@@ -611,10 +611,7 @@ def from_cupy(array) -> ViewType:
         layout = Layout.LayoutRight
 
     memory_space: MemorySpace
-    if km.get_gpu_framework() == "Cuda":
-        memory_space = MemorySpace.CudaSpace
-    elif km.get_gpu_framework() == "HIP":
-        memory_space = MemorySpace.HIPSpace
+    memory_space = MemorySpace.CudaSpace
 
     return from_numpy(np_array, memory_space, layout)

```

This almost passes all the `test_views.py` tests, except for GPU/`bool` stuff, which maybe isn't surprising (`1 failed, 45 passed in 34.82`):

```
________________________________________________________________________________________________________________________________ TestViews.test_real _________________________________________________________________________________________________________________________________

self = <tests.test_views.TestViews testMethod=test_real>

    @pytest.mark.skipif(not HAS_CUDA,
                        reason="CUDA/cupy not available")
    def test_real(self):
        pk.set_default_precision(pk.int32)
        view: pk.View1d = pk.View([self.threads])
    
        self.assertTrue(view.dtype is pk.DataType.int32)
        self.assertTrue(pk.View._get_dtype_name(str(type(view.array))) == "int32")
    
        f = RealViewTestFunctor(view)
        w = RealViewTestWorkload(self.threads, view)
>       pk.parallel_for(self.threads, f.pfor)

tests/test_views.py:313: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
pykokkos/interface/parallel_dispatch.py:167: in parallel_for
    func, args = runtime_singleton.runtime.run_workunit(
pykokkos/core/runtime.py:97: in run_workunit
    members: Optional[PyKokkosMembers] = self.precompile_workunit(workunit,policy.space)
pykokkos/core/runtime.py:68: in precompile_workunit
    members: Optional[PyKokkosMembers] = self.compiler.compile_object(module_setup, space, km.is_uvm_enabled())
pykokkos/core/compiler.py:146: in compile_object
    self.compile_entity(module_setup.main, module_setup, entity, parser.get_classtypes(), space, force_uvm, members)
pykokkos/core/compiler.py:187: in compile_entity
    functor, bindings = translator.translate(entity, classtypes)
pykokkos/core/translators/static.py:71: in translate
    bindings: List[str] = self.generate_bindings(entity, functor_name, source, workunits)
pykokkos/core/translators/static.py:261: in generate_bindings
    bindings = bind_workunits(functor_name, self.pk_members, workunits, self.module_file)
pykokkos/core/translators/bindings.py:468: in bind_workunits
    w, b = bind_workunits_single(functor, members, workunits, d)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

functor = 'pk_functor_RealViewTestFunctor', members = <pykokkos.core.translators.members.PyKokkosMembers object at 0x7fdcc0135ed0>
workunits = {<pykokkos.core.cppast.expr.DeclRefExpr object at 0x7fdcc0135630>: ('for', <pykokkos.core.cppast.decl.MethodDecl object at 0x7fdcc01365f0>)}, precision = <DataType.bool: <class 'numpy.bool_'>>

    def bind_workunits_single(
        functor: str,
        members: PyKokkosMembers,
        workunits: Dict[cppast.DeclRefExpr, Tuple[str, cppast.MethodDecl]],
        precision: Optional[DataType]
    ) -> Tuple[List[str], List[str]]:
        """
        Generates the bindings for a group of workunits. Each workunit is
        called inside a kernel, and each kernel is called from a wrapper
        that is bound with pybind.
    
        :param functor: the functor class name
        :param members: an object containing the fields and views
        :param workunits: a dictionary mapping form workunit name to a tuple of operation type and source
        :param precision: the precision for which to generate a binding
        :returns: a tuple of lists of strings of representing the wrapper names, and the kernels and wrappers
        """
    
        bindings: List[str] = []
        wrappers: List[str] = []
    
        real: Optional[str] = None
        if precision is not None:
>           real = visitors_util.view_dtypes[precision.name].value
E           KeyError: 'bool'

pykokkos/core/translators/bindings.py:421: KeyError

```

`nvidia-smi` shows this, for what I think is an old 1080Ti or something:

```
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 515.65.01    Driver Version: 515.65.01    CUDA Version: 11.7     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  NVIDIA GeForce ...  Off  | 00000000:B3:00.0  On |                  N/A |
|  0%   57C    P0    68W / 275W |    669MiB / 11264MiB |      2%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+

```

--- a/pykokkos/kokkos_manager/__init__.py
+++ b/pykokkos/kokkos_manager/__init__.py
@@ -204,20 +204,29 @@ if pk_kokkos_version is not None:
     except ValueError:
         print(f"WARNING: PK_KOKKOS_INTERFACE value '{pk_kokkos_version}' is invalid; reverting to {get_kokkos_version()}")
 
-try:
-    # Import multiple kokkos libs to support multiple devices per
-    # process. This assumes that there are modules named f"gpu{id}"
-    # that can be imported.
-    import atexit
-    import cupy as cp
-    import importlib
-    import sys
+# NOTE: multiple GPU support is almost certainly
+# broken, we can't just assume that there are modules
+# named gpu0, gpu1, and so on...
 
+# Import multiple kokkos libs to support multiple devices per
+# process. This assumes that there are modules named f"gpu{id}"
+# that can be imported.
+import atexit
+import importlib
+import sys
+
+try:
+    import cupy as cp
     NUM_CUDA_GPUS: int = cp.cuda.runtime.getDeviceCount()
     KOKKOS_LIBS: List[str] = [f"gpu{id}" for id in range(NUM_CUDA_GPUS)]
+except ImportError:
+    NUM_CUDA_GPUS = 0
+    KOKKOS_LIBS = []
 
-    KOKKOS_LIB_INSTANCES: List = []
-    for id, lib in enumerate(KOKKOS_LIBS):
+
+KOKKOS_LIB_INSTANCES: List = []
+for id, lib in enumerate(KOKKOS_LIBS):
+    try:
         module = importlib.import_module(lib)
         KOKKOS_LIB_INSTANCES.append(module)
 
@@ -228,16 +237,16 @@ try:
         module.initialize()
         atexit.register(module.finalize)
         sys.argv.pop()
+    except ModuleNotFoundError:
+        pass
 
+if len(KOKKOS_LIB_INSTANCES) > 1:
     CONSTANTS["MULTI_GPU"] = True
     CONSTANTS["NUM_GPUS"] = NUM_CUDA_GPUS
     CONSTANTS["KOKKOS_GPU_MODULE_LIST"] = KOKKOS_LIB_INSTANCES
     CONSTANTS["KOKKOS_GPU_MODULE"] = KOKKOS_LIB_INSTANCES[0]
 
-    if kokkos.get_device_available("Cuda"):
-        CONSTANTS["GPU_BACKEND"] = "Cuda"
-    elif kokkos.get_device_available("HIP"):
-        CONSTANTS["GPU_BACKEND"] = "HIP"
-
-except Exception:
-    pass
+if kokkos.get_device_available("Cuda"):
+    CONSTANTS["GPU_BACKEND"] = "Cuda"
+elif kokkos.get_device_available("HIP"):
+    CONSTANTS["GPU_BACKEND"] = "HIP"

--- a/tests/test_views.py
+++ b/tests/test_views.py
@@ -301,6 +301,8 @@ class TestViews(unittest.TestCase):
 
     @pytest.mark.skipif(not HAS_CUDA,
                         reason="CUDA/cupy not available")
+    @pytest.mark.xfail(HAS_CUDA,
+                       reason="bool not supported with CUDA/GPUs yet")
     def test_real(self):
         pk.set_default_precision(pk.int32)
         view: pk.View1d = pk.View([self.threads])

